pkg/mounter/safe_mounter_windows.go

```diff
@@ -118,8 +126,24 @@ func (mounter *csiProxyMounter) SMBMount(source, target, fsType string, mountOpt
 
 func (mounter *csiProxyMounter) SMBUnmount(target string) error {
 	klog.V(4).Infof("SMBUnmount: local path: %s", target)
-	// TODO: We need to remove the SMB mapping. The change to remove the
-	// directory brings the CSI code in parity with the in-tree.
+
+	parentDir := filepath.Dir(target)
+	if buffer, err := os.ReadFile(parentDir + string(os.PathSeparator) + "remote.txt"); err != nil {
```

Comment:
what's this `ReadFile` operation for?

Reply:
```
	SMBMount(source, target, fsType string, mountOptions, sensitiveMountOptions []string) error
	SMBUnmount(target string) error
```

`SMBUnmount` dont have `source` parameter

Reply:
but `remote.txt` won't exist, right? What about use `GetSmbGlobalMapping` to get the source?

Reply:
`remote.txt` will exist because we create it in SMBMount functions.
We can't use `GetSmbGlobalMapping` becase we don't have one-to-one relations between global mapping and remote path. I described this case in `cmb-proxy` PR.

Reply:
If `remote.txt` doesn't exist we will ignore it and don't call `RemoveSmbGlobalMapping`.

Reply:
In `NodeUnpublishVolume`, we only need to remove symbolic link, and in `NodeUnstageVolume`, we could use `GetSmbGlobalMapping` to get remote path, and remove GlobalMapping.

Reply:
I agree with you:
- about `NodeUnpublishVolume`. So, I changed `CleanupSMBMountPoint` to `CleanupMountPoint`.
- about `GetSmbGlobalMapping`. We need to know `RemotePath`. So, we save it to `remote.txt` when we call `SMBMount`. And when we call 'SMBUnmount' we load `RemotePath` from `remote.txt` and call `RemoteSmbGlobalMapping` via proxy.

Reply:
`Get-SmbGlobalMapping` would return `RemotePath`:
https://docs.microsoft.com/en-us/powershell/module/smbshare/get-smbglobalmapping?view=windowsserver2022-ps#example-1-retrieve-smb-client-directory-global-mappings

There is a problem that if two PVs reference same remote path, and in one PV `NodeUnstageVolume`, the remote path would be removed, that would break the other PV using same remote path, we don't have a way to check the ref count between symbolic link and remote path, it could be n:1 mapping, that's reason why we don't remove SmbGlobalMapping in `NodeUnstageVolume` now:
https://github.com/kubernetes-csi/csi-proxy/blob/26b63f840692b3145ef4005652a1b9b94146b14e/pkg/server/smb/server.go#L46-L57

Reply:
We can't use `Get-SmbGlobalMapping` to obtain `RemotePath` because `LocalPath` it is drive latter. It is not our local path that we use to make a link.

About problem with two PVs that reference same file share (not a remote path) I used counters in csi-proxy repo.

Reply:
Moreover, when we call `New-SmbGlobalMapping` we don't specify local path at all because it doesn't matter. This call required only for authentication.

Reply:
I implemented ref counter between symbolic link and remote path in csi-proxy.

Reply:
Last commit removes using of `remote.txt`. I think we can close the conversation.

Reply:
thanks, that's quite neat. While I still would like to make sure https://github.com/kubernetes-csi/csi-proxy/pull/210 is fixed first since without csi-proxy support and when this PR merged, it would remove global mapping in every PV unmount, that would cause big problem.

Reply:
I agree with you. It might be a problem. But as I understand, the current csi-proxy doesn't work correctly in this case too even without this PR.

Reply:
without this PR, the driver won't remove the global mapping

Reply:
yep. but without csi-proxy PR it can't create more than one mapping for the smb share.

Reply:
without https://github.com/kubernetes-csi/csi-proxy/pull/210 and this PR, user can only creates one global mapping for two PVs with same smb file share, right? If so, removing the shared global mapping would cause another PV mount failure, correct? I would like to figure it out first, thanks.

I am thinking add a switch for this "removing smb global mapping", that may be more flexible.
Sorry for the back-and-forth questions, I would like to make it quite clear since it may potential break sth.

Reply:
Add a switch - good idea. It can close all issues with backward compatibility. I will try to do it.

Reply:
I've added the switch via use env variable 'ENABLE_REMOVING_SMB_GLOBAL_MAPPING'. By default, it is empty and global mapping isn't called. To enable this feature, it needs to set value 'true' to the variable. We can do it in YAML for DaemonSet.

I didn't use command line switches to avoid adding a difference between windows and linux. This trick is needed only for windows.

Reply:
@andyzhangx have you seen last changes?

Reply:
Ref counter has been moved to this repo. So 'ENABLE_REMOVING_SMB_GLOBAL_MAPPING' is not needed now.
Currently we will call Remove-SmbGlobalMapping when it necessary and it won't break something with old csiproxy.